### PR TITLE
Add Prettier for linting frontend code

### LINKED ISSUE #2466: Use Prettier
@stephanwlee and I want to run [Prettier][prettier] on TensorBoard.

Prettier is a code formatter for JavaScript, TypeScript, HTML, and CSS
(among others). Its approach is different than that of most formatting
linters, like ESLint: instead of identifying “bad patterns” and
suggesting peephole fixes, it simply parses the entire file into a
concrete syntax tree and re-emits the tree with a pretty-printing
algorithm. This “uncompromising” nature makes it a pleasure to work
with: it’s pretty much entirely insensitive to however the input was
already formatted, and will give you entirely predictable results.

(It’s like Go’s `gofmt`, or Python’s [`black`][black].)

Anecdotally, I noticed that my process of writing code changed fairly
significantly when using Prettier. I can write hideous-looking one-line
monsters, then simply save the file and have it automatically formatted
correctly. (And if it doesn’t reformat on save then I know that I have a
syntax error!) Unlike with ESLint and friends, I can _actually_
completely forget about the formatting.

## Why?

Most importantly, you don’t have to think about formatting while
writing, reviewing, or reading code.

Consistent style also makes diffs smaller, and improves greppability
(e.g., you don’t have to account for two quote forms).

### Style

Prettier’s code style uses a bit more whitespace than our existing code
style. Running Prettier on our codebase increases the total LOC in JS,
TS, HTML, and CSS files by 18%.

Prettier also indents TypeScript namespaces, while we historically have
not, so this adds ~1 level of indentation to all TypeScript files.

This is all totally fine, imho; just wanted to mention it.

### One-shot versus per-file incremental

The diff is fairly large:

```
$ git ls-files -z '*.js' '*.ts' '*.html' '*.css' | xargs -0 wc -l | tail -n 1
  87771 total
$ prettier --write 'tensorboard/**/*.'{js,ts,html,css} >/dev/null 2>/dev/null
$ git ls-files -z '*.js' '*.ts' '*.html' '*.css' | xargs -0 wc -l | tail -n 1
 103474 total
$ git diff --shortstat
 367 files changed, 70988 insertions(+), 55286 deletions(-)
```

Essentially every file is modified, as should be expected. Some tools
will choke when trying to show a diff for a commit this large (e.g.,
[vim-fugitive]). There’s a valid argument that we shouldn’t design
around those broken tools. But Prettier also supports an incremental
per-file option:

```
  --require-pragma         Require either '@prettier' or '@format' to be present in the file's first docblock comment
                           in order for it to be formatted.
                           Defaults to false.
```

This would enable us to reformat fewer files at a time, while still
protecting against backsliding. We could adopt a soft policy that we
reformat files before modifying them, for instance.

Which should we choose?

## Steps

Here’s what we need to do to make this happen.

  - [x] Make sure that all code is syntactically valid, so that Prettier
    can parse it. Done in #2423.
  - [ ] Check in a `.prettierconfig`. (We like [this one][pc].)
  - Either:
      - [ ] Add `prettier -l 'tensorboard/**/*.'{js,ts,html,css}` to CI,
            and reformat all code.
  - Or:
      - [ ] Add `prettier --require-pragma -l 'tensorboard/**/*.'{js,ts,html,css}`
        to CI.
      - [ ] Adopt a policy of adding `@format` to all new files.
      - [ ] Adopt a soft policy of generally formatting files, modules,
        or packages (choose one?) before working on them.
  - [ ] Update `CONTRIBUTING.md` to show how to configure format-on-save.

[black]: https://github.com/psf/black
[pc]: https://github.com/sourcecred/sourcecred/blob/d5a1ca30b4d1551fdc3bf81b488aa8ae4a37c1d3/.prettierrc.json
[prettier]: https://prettier.io/playground/
[vim-fugitive]: https://github.com/tpope/vim-fugitive

### PR DESCRIPTION
**Pull request by wchargin**
Fixes #2466. See commit messages for discussion of configuration choices
and the script used to generate the reformatting commits.

### CODE REVIEW

**Comment by wchargin** [comment]
CLAs are fine because I wrote or generated all of these commits.

The author is intentionally tensorboard-gardener@google.com.

**Comment by wchargin** [comment]
Travis failures are unrelated: the `tf-nightly` builds are installing
versions of `tf-nightly` from 2019-07-30 (???), which was the one
nightly that was switched over to the V2 APIs, and so tests are failing
(e.g., because contrib is gone).

No idea why Travis is trying to install those old versions; I’ll clear
the build caches and retry.

(By this point in the build, Prettier has already passed.)

**Comment by wchargin** [comment]
Tests pass after cherry-picking #2494.

File: tensorboard/components/tf_backend/requestManager.ts [hunk 1]
```diff
@@ -13,287 +13,302 @@ See the License for the specific language governing permissions and
...
-    return this.requestWithOptions(url, requestOptions);
-  }
+    /**
+     * Gives a promise that loads assets from given url (respects queuing). If
+     * postData is provided, this request will use POST, not GET. This is an
+     * object mapping POST keys to string values.
+     */
+    public request(
+      url: string,
+      postData?: {[key: string]: string}
```
Comment (stephanwlee) [inline tensorboard/components/tf_backend/requestManager.ts]: Is it impossible to get trailing comma here too?
Rationale is the same as other trailing commas: I want git-diff to be minimal when adding/removing arguments.

File: tensorboard/components/tf_dashboard_common/dashboard-style.html [hunk 2]
```diff
@@ -54,8 +54,9 @@
       }
 
       .sidebar-section {
-        border-top: solid 1px rgba(0, 0, 0, .12);
-        padding: var(--sidebar-vertical-padding) 0 var(--sidebar-vertical-padding) var(--sidebar-left-padding);
+        border-top: solid 1px rgba(0, 0, 0, 0.12);
```
Comment (stephanwlee) [inline tensorboard/components/tf_dashboard_common/dashboard-style.html]: This change violates: https://google.github.io/styleguide/htmlcssguide.html#Leading_0s
Honestly, I think it is job of modernizer to get rid of "0" for shaving few bytes off of CSS payload so I am okay with whatever if it is hard to configure the prettier.

Comment (wchargin) [reply tensorboard/components/tf_dashboard_common/dashboard-style.html]: Agreed; the leading 0 makes it easier to read, minifcation can be done
by a compiler, and Prettier has no configuration option here, so I’ll
leave this alone.

File: tensorboard/components/tf_backend/requestManager.ts (hunk 1 above)
Comment (wchargin) [reply tensorboard/components/tf_backend/requestManager.ts]: I wish, but our version of TypeScript doesn’t support that:

```
ERROR: ${PWD}/tensorboard/plugins/histogram/tf_histogram_dashboard/BUILD:7:1: Compiling 34 TypeScript files //tensorboard/plugins/histogram/tf_histogram_dashboard:tf_histogram_dashboard failed (Exit 1) execrooter failed: error executing command bazel-out/host/bin/tensorboard/scripts/execrooter bazel-out/k8-fastbuild/bin/tensorboard/plugins/histogram/tf_histogram_dashboard/tf_histogram_dashboard-tsc-execroot.json ... (remaining 16 argument(s) skipped)

Use --sandbox_debug to see verbose messages from the sandbox
../../../../../../../../../../../../../../tmp/tmp_VP6Gz/tf-histogram-dashboard/histogramCore.ts(23,11): error TS1009: Trailing comma not allowed.
../../../../../../../../../../../../../../tmp/tmp_VP6Gz/tf-histogram-dashboard/histogramCore.ts(29,26): error TS1009: Trailing comma not allowed.
```

See:

  - <https://github.com/prettier/prettier/issues/1820>
  - <https://github.com/microsoft/TypeScript/issues/16152>

If at some point we upgrade our toolchains, we can flip this flag and
re-Prettify; the diff is comparatively small (1800 lines).

Comment (stephanwlee) [reply tensorboard/components/tf_backend/requestManager.ts]: Wow, didn't realize our version of tsc didn't support this. Okay, LGTM.

**Comment by wchargin** [comment]
> remove the [testing...] bit

This will be rebased, but sure. The rebase will also likely get clabot
to complain again; I’ll take care of that when it happens. :-)

**Comment by nfelt** [comment]
Were you thinking we'd rebase-and-merge this versus squash-and-merge?

Rebasing 100+ commits onto master is going to leave several useless screenfulls of commit lines in our commit history.  I know we were talking about keeping commit size down, but is there some way we could dial this into the range of say <20 somewhat larger commits so it's not quite as much of an impact to the commit history?  Or just squash it in after all?

**Comment by wchargin** [comment]
> Were you thinking we'd rebase-and-merge this versus squash-and-merge?

Yes, [per discussion][rebase].

[rebase]: https://github.com/tensorflow/tensorboard/issues/2466#issuecomment-515635336

> is there some way we could dial this into the range of say <20
> somewhat larger commits

Sure; we can group by parent directory to reduce to 21 commits, the
largest of which has a 17036 insertions. Viewing a commit of this size
in `vim-fugitive` is moderately slow (about 1 second), but much faster
than (e.g.) the initial TensorFlow commit.

<details>
<summary>Breakdown of <code>--shortstat</code>s</summary>

```
reformat tensorboard/plugins/interactive_inference/witwidget/notebook/jupyter/js:  4 files changed, 111 insertions(+), 98 deletions(-)
skip tensorboard/plugins/interactive_inference/witwidget/notebook/jupyter (covered by subtrees)
reformat tensorboard/plugins/example/tensorboard_plugin_example:  1 file changed, 17 insertions(+), 17 deletions(-)
reformat tensorboard/plugins/interactive_inference:  15 files changed, 7437 insertions(+), 5048 deletions(-)
reformat tensorboard/plugins/profile/memory_viewer:  15 files changed, 1108 insertions(+), 960 deletions(-)
reformat tensorboard/plugins/profile/pod_viewer:  8 files changed, 1350 insertions(+), 1150 deletions(-)
reformat tensorboard/plugins/custom_scalar:  5 files changed, 716 insertions(+), 600 deletions(-)
reformat tensorboard/plugins/distribution:  5 files changed, 757 insertions(+), 270 deletions(-)
reformat tensorboard/plugins/histogram:  8 files changed, 1784 insertions(+), 744 deletions(-)
reformat tensorboard/plugins/projector:  57 files changed, 11616 insertions(+), 10256 deletions(-)
reformat tensorboard/plugins/beholder:  3 files changed, 436 insertions(+), 401 deletions(-)
reformat tensorboard/plugins/debugger:  14 files changed, 3154 insertions(+), 2678 deletions(-)
reformat tensorboard/plugins/pr_curve:  3 files changed, 371 insertions(+), 311 deletions(-)
reformat tensorboard/plugins/hparams:  26 files changed, 3863 insertions(+), 3215 deletions(-)
reformat tensorboard/plugins/profile:  12 files changed, 3089 insertions(+), 2593 deletions(-)
reformat tensorboard/plugins/scalar:  3 files changed, 303 insertions(+), 251 deletions(-)
reformat tensorboard/plugins/audio:  3 files changed, 115 insertions(+), 101 deletions(-)
reformat tensorboard/plugins/graph:  58 files changed, 15876 insertions(+), 14003 deletions(-)
reformat tensorboard/plugins/image:  2 files changed, 137 insertions(+), 94 deletions(-)
reformat tensorboard/plugins/mesh:  3 files changed, 409 insertions(+), 343 deletions(-)
reformat tensorboard/plugins/text:  2 files changed, 67 insertions(+), 46 deletions(-)
reformat tensorboard/components:  122 files changed, 17036 insertions(+), 10875 deletions(-)
skip tensorboard/plugins (covered by subtrees)
skip tensorboard (covered by subtrees)
```

</details>

**Comment by wchargin** [comment]
CLAs are fine because I wrote or generated all of these commits.

The author is intentionally tensorboard-gardener@google.com.

**Pull request merged**